File: src/compute/scala_transformer/OmniTransformer.scala
// OmniTransformer.scala — Scala Transformer Implementation
// Inspired by: Memformer + RQ-Transformer architecture
// Layer: Compute / Scala
//
// Type-safe transformer with pattern matching, immutable state,
// and monadic error handling in Scala's type system.

package omni.compute.transformer

import scala.math.{sqrt, exp, log, tanh, Pi}
import scala.collection.immutable.Vector

sealed trait TransformerError
case class DimensionMismatch(expected: Int, got: Int) extends TransformerError
case class InvalidConfig(message: String) extends TransformerError
case class NumericalInstability(location: String) extends TransformerError

case class TransformerConfig(
  dim: Int = 512,
  heads: Int = 8,
  ffMult: Int = 4,
  depth: Int = 6,
  dropout: Double = 0.1,
  maxSeqLen: Int = 2048,
  vocabSize: Int = 32000
) {
  require(dim % heads == 0, s"dim ($dim) must be divisible by heads ($heads)")
  val headDim: Int = dim / heads
  val ffDim: Int = dim * ffMult
}

case class Tensor(data: Vector[Vector[Double]], rows: Int, cols: Int) {
  require(data.length == rows, s"Expected $rows rows, got ${data.length}")

  def apply(row: Int, col: Int): Double = data(row)(col)

  def +(other: Tensor): Either[TransformerError, Tensor] = {
    if (rows != other.rows || cols != other.cols)
      Left(DimensionMismatch(rows * cols, other.rows * other.cols))
    else {
      val result = data.zip(other.data).map { case (a, b) =>
        a.zip(b).map { case (x, y) => x + y }
      }
      Right(Tensor(result, rows, cols))
    }
  }

  def matMul(other: Tensor): Either[TransformerError, Tensor] = {
    if (cols != other.rows)
      Left(DimensionMismatch(cols, other.rows))
    else {
      val otherT = other.transpose
      val result = data.map { row =>
        otherT.data.map { col =>
          row.zip(col).map { case (a, b) => a * b }.sum
        }
      }
      Right(Tensor(result, rows, other.cols))
    }
  }

  def transpose: Tensor = {
    val transposed = (0 until cols).map { j =>
      (0 until rows).map { i => data(i)(j) }.toVector
    }.toVector
    Tensor(transposed, cols, rows)
  }

  def scale(factor: Double): Tensor = {
    Tensor(data.map(_.map(_ * factor)), rows, cols)
  }
}

object Tensor {
  def zeros(rows: Int, cols: Int): Tensor =
    Tensor(Vector.fill(rows)(Vector.fill(cols)(0.0)), rows, cols)

  def ones(rows: Int, cols: Int): Tensor =
    Tensor(Vector.fill(rows)(Vector.fill(cols)(1.0)), rows, cols)
}

object Activations {
  def gelu(x: Double): Double =
    0.5 * x * (1.0 + tanh(sqrt(2.0 / Pi) * (x + 0.044715 * x * x * x)))

  def softmax(xs: Vector[Double]): Vector[Double] = {
    val maxVal = xs.max
    val exps = xs.map(x => exp(x - maxVal))
    val sumExps = exps.sum
    exps.map(_ / sumExps)
  }
}

object LayerNorm {
  def apply(xs: Vector[Double], eps: Double = 1e-6): Vector[Double] = {
    val n = xs.length.toDouble
    val mean = xs.sum / n
    val variance = xs.map(x => (x - mean) * (x - mean)).sum / n
    val std = sqrt(variance + eps)
    xs.map(x => (x - mean) / std)
  }

  def applySeq(tensor: Tensor): Tensor = {
    val normed = tensor.data.map(LayerNorm.apply(_))
    Tensor(normed, tensor.rows, tensor.cols)
  }
}

object Attention {
  def scaledDotProduct(
    queries: Tensor, keys: Tensor, values: Tensor
  ): Either[TransformerError, (Tensor, Tensor)] = {
    val scale = 1.0 / sqrt(queries.cols.toDouble)
    val scaledQ = queries.scale(scale)

    for {
      scores <- scaledQ.matMul(keys.transpose)
      weights = Tensor(
        scores.data.map(row => Activations.softmax(row)),
        scores.rows, scores.cols
      )
      attended <- weights.matMul(values)
    } yield (attended, weights)
  }

  def multiHead(
    config: TransformerConfig, query: Tensor, key: Tensor, value: Tensor
  ): Either[TransformerError, Tensor] = {
    val headDim = config.headDim
    val numHeads = config.heads

    val headResults = (0 until numHeads).map { h =>
      val qHead = Tensor(
        query.data.map(row => row.slice(h * headDim, (h + 1) * headDim)),
        query.rows, headDim
      )
      val kHead = Tensor(
        key.data.map(row => row.slice(h * headDim, (h + 1) * headDim)),
        key.rows, headDim
      )
      val vHead = Tensor(
        value.data.map(row => row.slice(h * headDim, (h + 1) * headDim)),
        value.rows, headDim
      )
      scaledDotProduct(qHead, kHead, vHead).map(_._1)
    }

    val collected = headResults.foldLeft(Right(Vector.empty[Tensor]): Either[TransformerError, Vector[Tensor]]) {
      case (Right(acc), Right(t)) => Right(acc :+ t)
      case (Left(e), _) => Left(e)
      case (_, Left(e)) => Left(e)
    }

    collected.map { heads =>
      val concatenated = (0 until query.rows).map { i =>
        heads.flatMap(_.data(i)).toVector
      }.toVector
      Tensor(concatenated, query.rows, config.dim)
    }
  }
}

object FeedForward {
  def apply(dim: Int, ffDim: Int, input: Tensor): Tensor = {
    val activated = input.data.map { row =>
      val projected = row.flatMap(v => Vector.fill(ffDim / dim)(v)).take(ffDim)
      val geluApplied = projected.map(Activations.gelu)
      geluApplied.grouped(ffDim / dim).map(_.head).toVector.take(dim)
    }
    Tensor(activated, input.rows, dim)
  }
}

case class TransformerLayer(config: TransformerConfig) {
  def forward(input: Tensor): Either[TransformerError, Tensor] = {
    val normed1 = LayerNorm.applySeq(input)

    for {
      attnOut <- Attention.multiHead(config, normed1, normed1, normed1)
      residual1 <- input + attnOut
      normed2 = LayerNorm.applySeq(residual1)
      ffOut = FeedForward(config.dim, config.ffDim, normed2)
      residual2 <- residual1 + ffOut
    } yield residual2
  }
}

case class OmniTransformerEncoder(config: TransformerConfig) {
  private val layers = (0 until config.depth).map(_ => TransformerLayer(config))

  def forward(input: Tensor): Either[TransformerError, Tensor] = {
    layers.foldLeft(Right(input): Either[TransformerError, Tensor]) {
      case (Right(current), layer) => layer.forward(current)
      case (left @ Left(_), _) => left
    }.map(LayerNorm.applySeq)
  }
}

File: src/compute/ocaml_transformer/omni_transformer.ml
(* @omni-layer Compute | @omni-lang OCaml | @omni-batch 18 | @omni-semester 16 *)
(* @omni-description OCaml type-safe transformer: algebraic data types for
   model config, functor-based attention, and composable layers. *)

module type ATTENTION = sig
  type t
  val create : int -> int -> t
  val forward : t -> float array array -> float array array
end

module ScaledDotProduct : ATTENTION = struct
  type t = { d_model : int; n_heads : int; scale : float }

  let create d_model n_heads =
    let head_dim = d_model / n_heads in
    { d_model; n_heads; scale = 1.0 /. sqrt (float_of_int head_dim) }

  let softmax xs =
    let mx = Array.fold_left max neg_infinity xs in
    let exps = Array.map (fun x -> exp (x -. mx)) xs in
    let s = Array.fold_left (+.) 0.0 exps +. 1e-10 in
    Array.map (fun e -> e /. s) exps

  let dot a b =
    let n = min (Array.length a) (Array.length b) in
    let rec aux acc i =
      if i >= n then acc
      else aux (acc +. a.(i) *. b.(i)) (i + 1)
    in aux 0.0 0

  let forward t x =
    let n = Array.length x in
    let scores = Array.init n (fun i ->
      Array.init n (fun j ->
        dot x.(i) x.(j) *. t.scale
      )
    ) in
    let attn = Array.map softmax scores in
    Array.init n (fun i ->
      let d = Array.length x.(0) in
      Array.init d (fun dd ->
        let sum = ref 0.0 in
        for j = 0 to n - 1 do
          sum := !sum +. attn.(i).(j) *. x.(j).(dd)
        done;
        !sum
      )
    )
end

let layer_norm ?(eps=1e-5) xs =
  let n = float_of_int (Array.length xs) in
  let mean = Array.fold_left (+.) 0.0 xs /. n in
  let var_ = Array.fold_left (fun acc x -> acc +. (x -. mean) ** 2.0) 0.0 xs /. n in
  let inv = 1.0 /. sqrt (var_ +. eps) in
  Array.map (fun x -> (x -. mean) *. inv) xs

type config = {
  d_model : int;
  n_heads : int;
  n_layers : int;
  ff_dim : int;
  vocab_size : int;
}

let default_config = {
  d_model = 768;
  n_heads = 12;
  n_layers = 6;
  ff_dim = 3072;
  vocab_size = 32000;
}

let transformer_block cfg x =
  let module Attn = ScaledDotProduct in
  let attn = Attn.create cfg.d_model cfg.n_heads in
  let attn_out = Attn.forward attn x in
  let residual1 = Array.init (Array.length x) (fun i ->
    layer_norm (Array.init (Array.length x.(i)) (fun d ->
      x.(i).(d) +. attn_out.(i).(d)
    ))
  ) in
  residual1

let transformer_encoder cfg input =
  let rec loop x i =
    if i >= cfg.n_layers then x
    else loop (transformer_block cfg x) (i + 1)
  in
  loop input 0

let classify cfg n_classes input =
  let encoded = transformer_encoder cfg input in
  let n = Array.length encoded in
  let d = Array.length encoded.(0) in
  let pooled = Array.init d (fun dd ->
    Array.fold_left (fun acc row -> acc +. row.(dd)) 0.0 encoded /. float_of_int n
  ) in
  let logits = Array.init n_classes (fun c ->
    let sum = ref 0.0 in
    for d = 0 to min 32 (Array.length pooled - 1) do
      sum := !sum +. pooled.(d) *. sin (float_of_int (c + 1) *. 0.001 *. float_of_int (d + 1))
    done;
    !sum
  ) in
  let mx = Array.fold_left max neg_infinity logits in
  let exps = Array.map (fun l -> exp (l -. mx)) logits in
  let s = Array.fold_left (+.) 0.0 exps +. 1e-10 in
  Array.map (fun e -> e /. s) exps
